### PR TITLE
Add train phase distillation

### PR DESCRIPTION
**Pull request by jinPrelude**
Training teacher agent in Pong environment.
train_phase_teacher_train_pong

Training student agent using train-phase states betweeun ep0 ~ ep100 gathered by teacher agent above.
student_training_pong

avg score of student agent : 16.3
[INFO] test 0 step: 1587 total score: 20
[INFO] test 1 step: 1974 total score: 21
[INFO] test 2 step: 1971 total score: 21
[INFO] test 3 step: 1974 total score: 21
[INFO] test 4 step: 1559 total score: 20
[INFO] test 5 step: 835 total score: -20
[INFO] test 6 step: 1592 total score: 20
[INFO] test 7 step: 1652 total score: 20
[INFO] test 8 step: 1593 total score: 20
[INFO] test 9 step: 1590 total score: 20

 Training teacher agent in Pong environment 2.
image

Training student agent using train-phase last 50,000 states gather by 440ep teacher agent above
 image

avg score of student agent : 20.6
[INFO] test 0 step: 2846 total score: 19
[INFO] test 1 step: 1974 total score: 21
[INFO] test 2 step: 1971 total score: 21
[INFO] test 3 step: 1974 total score: 21
[INFO] test 4 step: 2846 total score: 19
[INFO] test 5 step: 1970 total score: 21
[INFO] test 6 step: 2851 total score: 19
[INFO] test 7 step: 2755 total score: 21
[INFO] test 8 step: 2751 total score: 21
[INFO] test 9 step: 2849 total score: 19

### CODE REVIEW

File: rl_algorithms/distillation/dqn_agent.py [hunk 1]
```diff
@@ -32,83 +35,162 @@ class DistillationDQN(DQNAgent):
...
-        self.curr_state = state
-        # epsilon greedy policy
-        # pylint: disable=comparison-with-callable
-        state = self._preprocess_state(state)
-        q_values = self.learner.dqn(state)
-
-        if not self.args.test and self.epsilon > np.random.random():
-            selected_action = np.array(self.env.action_space.sample())
+
+        if self.args.teacher:
```
Comment (MrSyee) [inline rl_algorithms/distillation/dqn_agent.py:82]: should do refactoring:
- move save logic to train method

File: rl_algorithms/distillation/dqn_agent.py [hunk 2]
```diff
@@ -134,30 +216,56 @@ def update_distillation(self) -> Tuple[torch.Tensor, ...]:
...
-        iter_1 = self.memory.buffer_size // self.hyper_params.batch_size
-        train_steps = iter_1 * self.hyper_params.epochs
-        print(
-            f"[INFO] Total epochs: {self.hyper_params.epochs}\t Train steps: {train_steps}"
-        )
-        n_epoch = 0
-        for steps in range(train_steps):
-            loss = self.update_distillation()
-
+        if self.args.student:
```
Comment (MrSyee) [inline rl_algorithms/distillation/dqn_agent.py:219]: should add comment
should modify description

File: rl_algorithms/distillation/dqn_agent.py [hunk 3]
```diff
@@ -134,30 +216,56 @@ def update_distillation(self) -> Tuple[torch.Tensor, ...]:
...
+                        f"Training {n_epoch} epochs, {steps} steps.. "
+                        + f"loss: {loss[0]}, avg_q_value: {loss[1]}"
+                    )
+                    self.learner.save_params(steps)
+                    n_epoch += 1
+                    self.memory.reset_dataloader()
+
+            self.learner.save_params(steps)
+
+        elif self.args.add_expert_q:
```
Comment (MrSyee) [inline rl_algorithms/distillation/dqn_agent.py:248]: should do refactoring
- combine this function with `train` method

File: run_pong_no_frameskip_v4.py [hunk 4]
```diff
@@ -97,10 +97,13 @@ def parse_args() -> argparse.Namespace:
         help="turn off framestack",
     )
     parser.add_argument(
-        "--distillation-buffer-path",
-        type=str,
-        default=None,
-        help="distillation buffer storage directory",
+        "--student", dest="student", action="store_true", help="train student",
+    )
+    parser.add_argument(
```
Comment (MrSyee) [inline run_pong_no_frameskip_v4.py:102]: should remove either `teacher` or `student` flag

File: rl_algorithms/distillation/dqn_agent.py (hunk 1 above)
Comment (jinPrelude) [reply rl_algorithms/distillation/dqn_agent.py]: resolved in e745842

File: rl_algorithms/common/buffer/distillation_buffer.py [hunk 5]
```diff
@@ -78,24 +69,30 @@ class DistillationDataset(Dataset):
...
         """
         self.buffer_path = buffer_path
         self.file_name_list = []
+        self.contain_q = 0
         for _dir in self.buffer_path:
             tmp = os.listdir(_dir)
             self.file_name_list += [_dir + "/" + x for x in tmp]
+            with open(self.file_name_list[-1], "rb") as f:
+                data = pickle.load(f)
+            self.contain_q = int(len(data) == 2)
```
Comment (MrSyee) [inline rl_algorithms/common/buffer/distillation_buffer.py:91]: should fix bug

Comment (MrSyee) [reply rl_algorithms/common/buffer/distillation_buffer.py:91]: should change boolean

File: rl_algorithms/distillation/dqn_agent.py [hunk 6]
```diff
@@ -8,11 +8,15 @@
 """
 
 import os
+import pickle
+import time
 from typing import Tuple
 
+from natsort import natsorted
```
Comment (MrSyee) [inline rl_algorithms/distillation/dqn_agent.py:15]: should change dictionary name like `000_dir`

File: configs/pong_no_frameskip_v4/distillation_dqn.py [hunk 7]
```diff
@@ -38,8 +40,9 @@
         ],
         # Distillation
         epochs=20,  # epoch of student training
-        buffer_size=int(50000),  # distillation buffer size
-        batch_size=32,  # distillation batch size
+        n_frame_from_last=int(
+            5e4
+        ),  # number of frames you want to save from the end of training
```
Comment (MrSyee) [inline configs/pong_no_frameskip_v4/distillation_dqn.py:45]: should reduce

**Review by MrSyee — APPROVED** [review]

**Pull request merged**